### ISSUE TITLE
ElasticSearch-Server upload with recent crawled data 

### ISSUE THREAD

**Issue by tilomi** [hochschule-darmstadt/openartbrowser#23]
**Is your feature request related to a problem? Please describe.**
New data was crawled between the 27.10.2019 and 30.10.2019. This data was crawled with the current version of ETL scripts. To test if the data can be uploaded on the production server i would recommend to install a local ElasticSearch server with the same configurations as the production server and test whether the upload works. In the next step we have to update existing indices and add new ones. There should be no website down time for this (?)

**Why do you want this feature? What goals should be achieved**
Update the openartbrowser data

**Describe the solution you'd like**
If possible we would adjust the elastic.py script according to our requirements.

**Describe acceptance criteria**
The new data is on the ElasticSearch-Server. There are no duplications or other data inconsistencies.

**Describe alternatives you've considered**
Upload to ElasticSearch with via web interface

**What effects does your proposed solution have**
More data on the ElasticSearch-Server

**Comment by tilomi** [comment]
I tested the extracted data on a local ElasticSearch installation. The elastic.py script worked for me i only had to set the encoding to utf-8. In the script the indexes are only created, to update the production server further small adjustments are needed.  I tested the local ElasticSearch server with a local frontend angular server with the right configuration on the ElasticSearch server everything worked as expected. So the json file seems to be created correctly.  I attached two screenshots from my local installation and one from the current production server to see the difference for the Mona Lisa artwork site (the only differences were in the details).

![Old Mona Lisa Data](https://user-images.githubusercontent.com/23573104/68070244-46585480-fd6c-11e9-865e-c198943d83ef.PNG)

![New Mona Lisa Data](https://user-images.githubusercontent.com/23573104/68070246-4f492600-fd6c-11e9-9312-a59019e7324f.PNG)

As you can see some motifs were added in the new data for the mona lisa

**Comment by tilomi** [comment]
To update the index we can copy the old index to another name as backup and create a new index with the old index's name. I think this would be the best solution because we can manually change back to the backup if there are bugs within the index.

**Comment by BernhardHumm** [comment]
Sehr gut, Herr Michel,

 

dann schalten Sie den neuen Index bitte live und informieren wieder . 

 

 

Viele Grüße

           Bernhard Humm

 

------------------------------------

Prof. Dr. Bernhard Humm

Geschäftsführender Direktor, Institut für angewandte Informatik Darmstadt (aiDa)

Vorsitzender Promotionsausschuss, Promotionszentrum Angewandte Informatik

h_da Hochschule Darmstadt - University of Applied Sciences, fbi Fachbereich Informatik 

Schöfferstr. 8b, D-64295 Darmstadt

Tel: +49 6151 16 38494, Fax: +49 6151 16 38935, Home Office: +49 6031 6714726

 <mailto:bernhard.humm@h-da.de> bernhard.humm@h-da.de, https://fbi.h-da.de/personen/personenseiten/bernhard-humm/

 

 

Von: Tilo Michel [mailto:notifications@github.com] 
Gesendet: Samstag, 2. November 2019 12:29
An: hochschule-darmstadt/openartbrowser <openartbrowser@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Betreff: Re: [hochschule-darmstadt/openartbrowser] ElasticSearch-Server upload with recent crawled data (#23)

 

I tested the extracted data on a local ElasticSearch installation. The elastic.py script worked for me i only had to set the encoding to utf-8. In the script the indexes are only created, to update the production server further small adjustments are needed. I tested the local ElasticSearch server with a local frontend angular server with the right configuration on the ElasticSearch server everything worked as expected. So the json file seems to be created correctly. I attached two screenshots from my local installation and one from the current production server to see the difference for the Mona Lisa artwork site (the only differences were in the details).

 <https://user-images.githubusercontent.com/23573104/68070244-46585480-fd6c-11e9-865e-c198943d83ef.PNG> 

 <https://user-images.githubusercontent.com/23573104/68070246-4f492600-fd6c-11e9-9312-a59019e7324f.PNG> 

As you can see some motifs were added in the new data for the mona lisa

**Comment by tilomi** [comment]
I changed the update script and tested the change from the old master_flat_rank.json to the new one. I still had a bug when updating (Updated entities like Mona Lisa weren't shown). If i figure out how to solve this i'll try it on the live server.

**Comment by jschurse** [comment]
> To update the index we can create a snapshot from the old index. And update it afterwards. I think this would be the best solution because we can manually change back to the backup if there are bugs within the index.

How long does the update take?
If you are creating a snapshot and modifying it, does the openartbrowser.org page still use the unmodified original data at this time?

Maybe we should run the update process on our (not yet available) staging system first.

**Comment by tilomi** [comment]
The update takes around half an hour. The index consists of documents which are artworks, artists and so on. They are for example accessed by openartbrowser.org/<index-name>/artists/<qId>. I'm now testing the update locally if it works there i'd assume it should work on the server aswell. I will create the snapshot before updating. If it fails for some other reasons i can't reproduce on a local server i would fallback to the snapshot. The documents to be updated are replaced step by step. The site should be callable while the update is ongoing. I'll test this locally if i found my bug.

**Comment by tilomi** [comment]
The problem was that i needed to get the old document ids from the index. I only searched by the qId but the qId is not unique because artworks like mona lisa are also motifs and motifs use the same qId. If i search by qId and type it should behave as expected and an update should be possible.

**Comment by tilomi** [comment]
On my local elastic search server everything is working according to the plan. I pushed my changes to a branch yesterday: [elasticsearch_upload](https://github.com/hochschule-darmstadt/openartbrowser/blob/feature/23_elasticsearch_upload/scripts/upload_to_elasticsearch/elastic.py) . I only have to figure out the file locations on the server and swap them in the script. To enable the ElasticSearch snapshot feature i have to change the configuration and restart the elastic search service. But this would be the only downtime needed from my side. While updating the index the website still works i tested this on my local setup. I also tested going back to a snapshot and it's very fast (around a second). I explained the problem about the elasticsearch configuration here: [Set backup folder for repository in ElasticSearch](https://github.com/hochschule-darmstadt/openartbrowser/blob/feature/23_elasticsearch_upload/scripts/upload_to_elasticsearch/elastic.py#L23). Before i change anything i'll wait for an ok.

**Comment by BernhardHumm** [comment]
Prima, Herr Michel,

 

und OK von meiner Seite

 

 

Viele Grüße

           Bernhard Humm

 

------------------------------------

Prof. Dr. Bernhard Humm

Geschäftsführender Direktor, Institut für angewandte Informatik Darmstadt (aiDa)

Vorsitzender Promotionsausschuss, Promotionszentrum Angewandte Informatik

h_da Hochschule Darmstadt - University of Applied Sciences, fbi Fachbereich Informatik 

Schöfferstr. 8b, D-64295 Darmstadt

Tel: +49 6151 16 38494, Fax: +49 6151 16 38935, Home Office: +49 6031 6714726

 <mailto:bernhard.humm@h-da.de> bernhard.humm@h-da.de, https://fbi.h-da.de/personen/personenseiten/bernhard-humm/

 

 

Von: Tilo Michel [mailto:notifications@github.com] 
Gesendet: Dienstag, 5. November 2019 10:11
An: hochschule-darmstadt/openartbrowser <openartbrowser@noreply.github.com>
Cc: Humm, Bernhard, Prof. Dr. <bernhard.humm@h-da.de>; Comment <comment@noreply.github.com>
Betreff: Re: [hochschule-darmstadt/openartbrowser] ElasticSearch-Server upload with recent crawled data (#23)

 

On my local elastic search server everything is working according to the plan. I pushed my changes to a branch yesterday: elasticsearch_upload <https://github.com/hochschule-darmstadt/openartbrowser/blob/feature/23_elasticsearch_upload/scripts/upload_to_elasticsearch/elastic.py>  . I only have to figure out the file locations on the server and swap them in the script. To enable the ElasticSearch snapshot feature i have to change the configuration and restart the elastic search service. But this would be the only downtime needed from my side. While updating the index the website still works i tested this on my local setup. I also tested going back to a snapshot and it's very fast (around a second). I explained the problem about the elasticsearch configuration here: Set backup folder for repository in ElasticSearch <https://github.com/hochschule-darmstadt/openartbrowser/blob/feature/23_elasticsearch_upload/scripts/upload_to_elasticsearch/elastic.py#L23> . Before i change anything i'll wait for an ok.
